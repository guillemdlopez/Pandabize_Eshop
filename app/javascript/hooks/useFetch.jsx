import { useEffect, useRef, useState } from "react";

const useFetch = (url) => {
  const isMounted = useRef(true);

  const [state, setState] = useState({
    data: null,
    loading: false,
    error: null,
  });

  useEffect(() => {
    return () => {
      //clean up process
      isMounted.current = false;
    };
  }, []);

  useEffect(() => {
    setState({ data: null, loading: true, error: null });
    fetch(url)
      .then((res) => {
        if (!res.ok) throw new Error(`There has been a problem ${res.status}`);

        return res.json();
      })
      .then((data) => {
        if (!data) throw new Error("No data was found!");
        if (isMounted.current) {
          setState({
            loading: false,
            error: null,
            data,
          });
        }
      })
      .catch((err) => {
        console.error(err);
      });
  }, [url]);

  return state;
};

export default useFetch;
