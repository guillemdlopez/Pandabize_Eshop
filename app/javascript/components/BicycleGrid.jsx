import React, { useContext } from "react";
import useFetch from "../hooks/useFetch";
import BicycleCard from "./BicycleCard";
import { userContext } from "../context/userContext";
import { API_BICYCLES_URL } from "../fixtures/variables";
import { server } from "../fixtures/config";

const BicycleGrid = () => {
  const { data, loading, error } = useFetch(`${server}${API_BICYCLES_URL}`);

  const { user } = useContext(userContext);
  return (
    <>
      <div className="banner">
        <div className="container">
          <h1 className="banner-greeting">
            Welcome back, <strong>{user}</strong>
          </h1>
        </div>
      </div>

      <div className="container">
        <div className="cards mb-5 mt-4 animate__animated animate__fadeIn">
          {data?.map((bicycle) => (
            <BicycleCard key={bicycle.id} {...bicycle} />
          ))}
        </div>
      </div>
    </>
  );
};

export default BicycleGrid;
