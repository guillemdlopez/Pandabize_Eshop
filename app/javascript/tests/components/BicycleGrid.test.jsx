import React from "react";
import { mount } from "enzyme";
import BicycleGrid from "../../components/BicycleGrid";
import { userContext } from "../../context/userContext";
import useFetch from "../../hooks/useFetch";
import { BrowserRouter as Route, Switch } from "react-router-dom";

jest.mock("../../hooks/useFetch");

describe("Testing the <BicycleGrid/> component", () => {
  const userInfo = {
    user: "Guillem",
    userId: 1,
  };

  useFetch.mockReturnValue({
    data: [
      {
        id: 1,
        name: "Prettiest Bicycle",
        description: "This is the prettiest bicycle in the world. Buy it!",
        photo: "image.jpg",
        customizations: [
          {
            id: 26,
            name: "Wheel size",
            value: "15 inches",
            price: 9.23,
            bicycle_id: 1,
          },
          {
            id: 148,
            name: "Wheel size",
            value: "16 inches",
            price: 6.15,
            bicycle_id: 1,
          },
        ],
      },
      {
        id: 2,
        name: "Coolest Bicycle",
        description: "This is the coolest bicycle in the world. Buy it!",
        photo: "image-2.jpg",
        customizations: [
          {
            id: 26,
            name: "Wheel size",
            value: "15 inches",
            price: 7.23,
            bicycle_id: 2,
          },
          {
            id: 148,
            name: "Wheel size",
            value: "16 inches",
            price: 5.15,
            bicycle_id: 2,
          },
        ],
      },
    ],
  });

  const wrapper = mount(
    <Route>
      <Switch>
        <userContext.Provider value={userInfo}>
          <BicycleGrid />
        </userContext.Provider>
      </Switch>
    </Route>
  );

  test("should render component correctly", () => {
    expect(wrapper).toMatchSnapshot();
  });

  test("should display a greeting message to the current user", () => {
    const { user } = userInfo;
    expect(wrapper.find("h1").text().trim()).toBe(`Welcome back, ${user}`);
    expect(wrapper.find("h1").text().trim()).not.toBe(`Welcome back, Antonio`);
  });

  test("should contain a banner", () => {
    expect(wrapper.find(".banner").exists()).toBe(true);
  });

  test("should contain a cards div", () => {
    expect(wrapper.find(".cards").exists()).toBe(true);
  });

  test("should display all the bicycles inside the cards div", () => {
    const cardsDiv = wrapper.find(".cards");
    expect(cardsDiv.find("a").length).toBe(2);
  });

  test("should render a BicycleCard for each object on the data's array", () => {
    expect(wrapper.find(".cards").children().length).toBe(2);
  });
});
