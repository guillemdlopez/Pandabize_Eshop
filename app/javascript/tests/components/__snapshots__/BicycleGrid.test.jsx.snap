// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing the <BicycleGrid/> component should render component correctly 1`] = `
Array [
  <div
    className="banner"
  >
    <div
      className="container"
    >
      <h1
        className="banner-greeting"
      >
        Welcome back, 
        <strong>
          Guillem
        </strong>
      </h1>
    </div>
  </div>,
  <div
    className="container"
  >
    <div
      className="cards mb-5 mt-4 animate__animated animate__fadeIn"
    >
      <a
        href="/bicycle/1"
        onClick={[Function]}
      >
        <div
          className="bike-card"
          data-bike-id={1}
        >
          <img
            className="bike-img"
            src="image.jpg"
          />
          <div
            className="bike-card-content"
          >
            <p
              className="bicycle-name"
            >
              <strong>
                Prettiest Bicycle
              </strong>
            </p>
            <p
              className="bicycle-description text-black-50"
            >
              This is the prettiest bicycle in the wor
              ...
            </p>
            <div
              className="d-flex mb-3"
            />
            <div
              className="d-flex align-items-center"
            >
              <p
                className="text-muted mb-0"
              >
                from
              </p>
              <p
                className="ml-2 lowest-price mb-0"
              >
                6.15
                 €
              </p>
            </div>
          </div>
        </div>
      </a>
      <a
        href="/bicycle/2"
        onClick={[Function]}
      >
        <div
          className="bike-card"
          data-bike-id={2}
        >
          <img
            className="bike-img"
            src="image-2.jpg"
          />
          <div
            className="bike-card-content"
          >
            <p
              className="bicycle-name"
            >
              <strong>
                Coolest Bicycle
              </strong>
            </p>
            <p
              className="bicycle-description text-black-50"
            >
              This is the coolest bicycle in the world
              ...
            </p>
            <div
              className="d-flex mb-3"
            />
            <div
              className="d-flex align-items-center"
            >
              <p
                className="text-muted mb-0"
              >
                from
              </p>
              <p
                className="ml-2 lowest-price mb-0"
              >
                5.15
                 €
              </p>
            </div>
          </div>
        </div>
      </a>
    </div>
  </div>,
]
`;
